src/core/lib/channel/channel_args.h

```diff
@@ -86,6 +88,32 @@ struct ChannelArgTypeTraits<
   };
 };
 
+// Specialization for shared_ptr
```

Comment:
I'd still rather get us to a place that we don't need to write `GetObjectRef<std::shared_ptr<T>>`...

As I see it the big reason we can't is that we can't go from `GetObjectRef<EventEngine>` to the conclusion that we need to wrap in a std::shared_ptr... so maybe we just give ourselves that hint:

```
template <typename T> struct WrapInSharedPtr : std::false_type {};
template <> struct WrapInSharedPtr<EventEngine> : std::true_type {};
```

and I suspect the rest of this shakes out so that SetObject/GetObject/GetObjectRef do as we'd like them to do.

Reply:
This PR adds general shared_ptr support to ChannelArgs, not just for EventEngine. I'd like to land it generally if possible. We could special-case EventEngine here if we wanted to add the same special cases for any other type we want to add as a shared_ptr.

I think the bigger blocker was `GetObject<T>` where we expect to return a raw pointer `p` in most cases, but `p->get()` when a shared_ptr is stored in channel_args. I can take another look.